source/common/quic/envoy_quic_server_session.cc

```diff
@@ -56,18 +56,31 @@ quic::QuicSpdyStream* EnvoyQuicServerSession::CreateIncomingStream(quic::QuicStr
   }
   auto stream = new EnvoyQuicServerStream(id, this, quic::BIDIRECTIONAL, codec_stats_.value(),
                                           http3_options_.value(), headers_with_underscores_action_);
+  initializeStream(stream);
+  return stream;
+}
+
+void EnvoyQuicServerSession::initializeStream(EnvoyQuicServerStream* stream) {
   ActivateStream(absl::WrapUnique(stream));
   if (aboveHighWatermark()) {
     stream->runHighWatermarkCallbacks();
   }
   setUpRequestDecoder(*stream);
-  return stream;
 }
 
-quic::QuicSpdyStream*
-EnvoyQuicServerSession::CreateIncomingStream(quic::PendingStream* /*pending*/) {
+quic::QuicSpdyStream* EnvoyQuicServerSession::CreateIncomingStream(quic::PendingStream* pending) {
   // Only client side server push stream should trigger this call.
```

Comment:
Sorry I'm a bit confused - if the client side sends a server push stream when it's not negotiated why don't we close the connection for it doing "illegal" things?  Should that be a TODO for quiche to fix upstream?

Reply:
Not just server push stream actually, instead, any client-initiated uni-directional stream with type byte other than the predesignated ones, QPack encode/decode streams, etc. QUICHE session should accept such stream as its extensions may wants to create such special streams. So it's not illegal for QUICHE.

I can close the connection in Envoy code if we don't have such use case in near future. Or update the comment to be more accurate?

Reply:
IIRC current QUIC support in Envoy is only for HTTP transactions which are always bi-directional.  Closing the connections seems appropriate until other use cases are added.

Reply:
https://github.com/envoyproxy/envoy/pull/17618 includes a QUICHE change that close the connection in this case. I will continue update this PR after that one is checked in.